<a href="https://colab.research.google.com/github/nrj177/SQL-Query-IMDB-dataset-analysis/blob/main/SQL_Query_IMDB_dataset_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

database: https://drive.google.com/file/d/1vlTeCUIfw_oxhpENKPJdcm1L34t6qdSg/view?usp=sharing

importing all libraries

In [36]:
#to perform mathematical operation
import numpy as np

#data manipulation tool
import pandas as pd

#data visualisation tool
import matplotlib.pyplot as plt

#data visualisation tool
import seaborn as sns

#server-less database
import sqlite3

#data cleaning
import re



1. Establishing the connection to the database

In [37]:
#storing the path of the database into a variable 'db'
db='/movies.sqlite'

#creating a database connection object
conn = sqlite3.connect(db)

#creating a cursor object
cur = conn.cursor()

In [38]:
# Fetch all the data of the movies table from the database
cur.execute('SELECT * FROM movies')

#creating a cursor object
movies = cur.fetchall()

In [39]:
#creating a DataFrame
movies_df = pd.DataFrame(movies, columns=['id','original_title','budget','popularity',
                                          'release_date','revenue','title','vote_average',
                                          'vote_count','overview','tagline','uid','director_id'] )


In [40]:
#displaying the dataframe
movies_df

id                            original_title  ...     uid  director_id
0     43597                                    Avatar  ...   19995         4762
1     43598  Pirates of the Caribbean: At World's End  ...     285         4763
2     43599                                   Spectre  ...  206647         4764
3     43600                     The Dark Knight Rises  ...   49026         4765
4     43601                               John Carter  ...   49529         4766
...     ...                                       ...  ...     ...          ...
4768  48395                               El Mariachi  ...    9367         5097
4769  48396                                 Newlyweds  ...   72766         6485
4770  48397                 Signed, Sealed, Delivered  ...  231617         7108
4771  48398                          Shanghai Calling  ...  126186         7109
4772  48399                         My Date with Drew  ...   25975         7110

[4773 rows x 13 columns]

In [41]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4773 entries, 0 to 4772
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              4773 non-null   int64  
 1   original_title  4773 non-null   object 
 2   budget          4773 non-null   int64  
 3   popularity      4773 non-null   int64  
 4   release_date    4773 non-null   object 
 5   revenue         4773 non-null   int64  
 6   title           4773 non-null   object 
 7   vote_average    4773 non-null   float64
 8   vote_count      4773 non-null   int64  
 9   overview        4770 non-null   object 
 10  tagline         3951 non-null   object 
 11  uid             4773 non-null   int64  
 12  director_id     4773 non-null   int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 484.9+ KB


In [42]:
# Fetch all the data of the Directors table from the database
cur.execute('SELECT * FROM directors')

#creating a cursor object
directors = cur.fetchall()

#creating dataframe for director table
directors_df = pd.DataFrame(directors, columns=['name','director_id','Gender','uid','department'])

In [43]:
directors_df.head()

name  director_id  Gender   uid department
0      James Cameron         4762       2  2710  Directing
1     Gore Verbinski         4763       2  1704  Directing
2         Sam Mendes         4764       2    39  Directing
3  Christopher Nolan         4765       2   525  Directing
4     Andrew Stanton         4766       2     7  Directing

In [44]:
directors_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2349 entries, 0 to 2348
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   name         2349 non-null   object
 1   director_id  2349 non-null   int64 
 2   Gender       2349 non-null   int64 
 3   uid          2349 non-null   int64 
 4   department   2349 non-null   object
dtypes: int64(3), object(2)
memory usage: 91.9+ KB


In [45]:
directors_df.describe()

director_id       Gender           uid
count  2349.000000  2349.000000  2.349000e+03
mean   5936.000000     1.404002  1.928479e+05
std     678.242213     0.879349  3.710910e+05
min    4762.000000     0.000000  1.000000e+00
25%    5349.000000     0.000000  1.776400e+04
50%    5936.000000     2.000000  5.822000e+04
75%    6523.000000     2.000000  1.096950e+05
max    7110.000000     2.000000  1.867383e+06

Questions

Q1. How many movies are present in the movies table?

In [46]:
movies_df.columns

Index(['id', 'original_title', 'budget', 'popularity', 'release_date',
       'revenue', 'title', 'vote_average', 'vote_count', 'overview', 'tagline',
       'uid', 'director_id'],
      dtype='object')

In [47]:
cur.execute('SELECT COUNT(original_title) FROM movies')
count = cur.fetchall()
print(f'The no:of movies present in movies table is {count[0]}')

The no:of movies present in movies table is (4773,)


* Conclusion: There are 4773 movies present in the table

Q2. Find these 3 directors: James Cameron, Luc Besson,John Woo

In [48]:
directors_df.columns

Index(['name', 'director_id', 'Gender', 'uid', 'department'], dtype='object')

In [49]:
cur.execute("SELECT * FROM directors where name IN  ('James Cameron','Luc Besson','John Woo')")
three_directors = cur.fetchall()
print(f'Here are the details of given directors: {three_directors}')


Here are the details of given directors: [('James Cameron', 4762, 2, 2710, 'Directing'), ('John Woo', 4893, 2, 11401, 'Directing'), ('Luc Besson', 4949, 2, 59, 'Directing')]


* Conclusion: All 3 directors found

Q3.Find all the directors with the name starting with 'Steven'

In [50]:
cur.execute("SELECT * FROM directors WHERE name LIKE 'Steven%'")
name = cur.fetchall()
print(f'Names starting with "steven" are: {name}')

Names starting with "steven" are: [('Steven Spielberg', 4799, 2, 488, 'Directing'), ('Steven Soderbergh', 4909, 2, 1884, 'Directing'), ('Steven Brill', 5013, 2, 32593, 'Directing'), ('Steven Zaillian', 5117, 2, 2260, 'Directing'), ('Steven Quale', 5216, 2, 93214, 'Directing'), ('Steven Seagal', 5221, 2, 23880, 'Directing'), ('Steven E. de Souza', 5390, 2, 1726, 'Directing'), ('Steven Shainberg', 5803, 2, 67795, 'Directing'), ('Steven R. Monroe', 6713, 2, 88039, 'Directing')]


* Conclusion: There are few names that is starting with "Steven"

Q4. Count all the female directors
Hint: Gender = 1

In [51]:
cur.execute("SELECT COUNT(*) FROM directors WHERE gender == '1' ")
fem_directors = cur.fetchall()
print(f'The No:of Female directors is: {fem_directors[0][0]}')


The No:of Female directors is: 150


* Conclusion: There are 150 female directors

Q5. Find the names of the 10th first female director

In [52]:
cur.execute("SELECT name FROM directors WHERE gender == '1'")
tenth = cur.fetchall()
print(f'The tenth female director is: {tenth[9][0]}')

The tenth female director is: Angelina Jolie


* Conclusion: Angelina Jolie is the 10th female director 

Q6. Find the top 3 most popular movies

In [53]:
movies_df.columns

Index(['id', 'original_title', 'budget', 'popularity', 'release_date',
       'revenue', 'title', 'vote_average', 'vote_count', 'overview', 'tagline',
       'uid', 'director_id'],
      dtype='object')

In [54]:
movies_df.popularity

0       150
1       139
2       107
3       112
4        43
       ... 
4768     14
4769      0
4770      1
4771      0
4772      1
Name: popularity, Length: 4773, dtype: int64

In [55]:
cur.execute("SELECT original_title FROM movies ORDER BY popularity DESC")
popular = cur.fetchall()
print(f'The top3 most popular movies are: {popular[0:3]}')

The top3 most popular movies are: [('Minions',), ('Interstellar',), ('Deadpool',)]


* Conclusion: Minions Interstellar & Deadpool are the most popular movies

Q7. Find the Top 3 most bankable(budget) movies

In [56]:
movies_df.columns

Index(['id', 'original_title', 'budget', 'popularity', 'release_date',
       'revenue', 'title', 'vote_average', 'vote_count', 'overview', 'tagline',
       'uid', 'director_id'],
      dtype='object')

In [57]:
cur.execute("SELECT original_title FROM movies ORDER BY budget DESC LIMIT 3")
budget = cur.fetchall()
print(f'Top 3 most bankable films are: {budget}')

Top 3 most bankable films are: [('Pirates of the Caribbean: On Stranger Tides',), ("Pirates of the Caribbean: At World's End",), ('Avengers: Age of Ultron',)]


* Conclusion: Pirates of the Caribbean: On Stranger Tides,Pirates of the Caribbean: At World's End & Avengers: Age of Ultron are the Top 3 most bankable movies

Q8. Find out the most awarded avg rated movie since jan 1st, 2000

In [58]:
cur.execute("SELECT original_title FROM movies WHERE release_date > '2000-01-01' ORDER BY vote_average DESC LIMIT 1")
avgRated = cur.fetchall()
print(f'The most awarded avg rated movie since jan 1st 2000 is: {avgRated[0][0]}')

The most awarded avg rated movie since jan 1st 2000 is: Sardaarji


* Conclusion: Sardaarji is the most awarded avg rated movie since jan 1st 2000

Q9. Find all the movies that has been directed by Brenda Chapman

In [59]:
cur.execute("SELECT original_title from movies LEFT JOIN directors ON movies.director_id = directors.id WHERE directors.name = 'Brenda Chapman'")
brenda = cur.fetchall()
print(f'Movie(s) that are/is directed by Brenda Chapman: {brenda[0][0]}')

Movie(s) that are/is directed by Brenda Chapman: Brave


* Conclusion: "Brave" is the only movie that is directed by Brenda Chapman

Q10. Find out the director who has made the most movies?

In [60]:
query = "SELECT name FROM directors JOIN movies ON directors.id = movies.director_id GROUP BY director_id ORDER BY COUNT(name) DESC LIMIT 1"
cur.execute(query)
dir = cur.fetchall()
print(f'The director with most movies: {dir[0][0]}')

The director with most movies: Steven Spielberg


* Conclusion: Steven Spielberg has directed the most no:of movies

Q11. Find out the director who is most bankable?

In [61]:
movies_df.columns

Index(['id', 'original_title', 'budget', 'popularity', 'release_date',
       'revenue', 'title', 'vote_average', 'vote_count', 'overview', 'tagline',
       'uid', 'director_id'],
      dtype='object')

In [62]:
query = "SELECT name FROM directors JOIN movies ON directors.id = movies.director_id GROUP BY director_id ORDER BY SUM(budget) DESC LIMIT 1"
cur.execute(query)
dir_mostBankable = cur.fetchall()
print(f'The director who is most bankable: {dir_mostBankable[0][0]}')

The director who is most bankable: Steven Spielberg


* Conclusion: Steven Spielberg is the most bankable director

Q12. Find out the top 10 highest budget making movies?

In [63]:
query = "SELECT original_title FROM movies ORDER BY budget DESC LIMIT 10"
cur.execute(query)
top_budget = cur.fetchall()
print(f'The top 10 highest budget movies are: {top_budget}')

The top 10 highest budget movies are: [('Pirates of the Caribbean: On Stranger Tides',), ("Pirates of the Caribbean: At World's End",), ('Avengers: Age of Ultron',), ('Superman Returns',), ('John Carter',), ('Tangled',), ('Spider-Man 3',), ('The Lone Ranger',), ('The Dark Knight Rises',), ('Harry Potter and the Half-Blood Prince',)]


* Conclusion: printed the top 10 highest budget movies

Q13. Find the top 10 revenue making movies

In [64]:
query = "SELECT original_title FROM movies ORDER BY revenue DESC LIMIT 10"
cur.execute(query)
top_revenue = cur.fetchall()
print(f'The top 10 highest revenue movies are: {top_revenue}')

The top 10 highest revenue movies are: [('Avatar',), ('Titanic',), ('The Avengers',), ('Jurassic World',), ('Furious 7',), ('Avengers: Age of Ultron',), ('Frozen',), ('Iron Man 3',), ('Minions',), ('Captain America: Civil War',)]


* Conclusion: printed the top 10 highest revenue movies

Q14. Find out the top 10 most popular movies with the highest rating

In [65]:
query = "SELECT original_title FROM movies ORDER BY vote_average DESC LIMIT 10"
cur.execute(query)
top_rating = cur.fetchall()
print(f'The top 10 most popular movies with the highest ratings are: {top_rating}')

The top 10 most popular movies with the highest ratings are: [('Stiff Upper Lips',), ('Dancer, Texas Pop. 81',), ('Sardaarji',), ("One Man's Hero",), ('The Shawshank Redemption',), ('There Goes My Baby',), ('The Prisoner of Zenda',), ('The Godfather',), ('Fight Club',), ("Schindler's List",)]


* Conclusion: Printed the top 10 most popular movies with the highest rating

Q15. Find the top10 directors with the no:of movies and revenue where revenue should be taken into account for doing  the analysis

In [71]:
query = "SELECT directors.name, COUNT(original_title), SUM(revenue) FROM directors JOIN movies ON directors.id = movies.director_id GROUP BY director_id ORDER BY SUM(revenue) DESC LIMIT 10"
cur.execute(query)
top_10 = cur.fetchall()
print(f'The top10 directors with the no:of movies and revenue are: {top_10}') 

The top10 directors with the no:of movies and revenue are: [('Steven Spielberg', 27, 9147393164), ('Peter Jackson', 9, 6498642820), ('James Cameron', 7, 5883569439), ('Michael Bay', 12, 5832524638), ('Christopher Nolan', 8, 4227483234), ('Chris Columbus', 11, 3725631503), ('Robert Zemeckis', 13, 3590622002), ('George Lucas', 5, 3339113893), ('Tim Burton', 14, 3337418241), ('Ridley Scott', 16, 3189557997)]


In [72]:
directors_df.columns

Index(['name', 'director_id', 'Gender', 'uid', 'department'], dtype='object')

In [68]:
movies_df.columns

Index(['id', 'original_title', 'budget', 'popularity', 'release_date',
       'revenue', 'title', 'vote_average', 'vote_count', 'overview', 'tagline',
       'uid', 'director_id'],
      dtype='object')

* Conclusion: Printed the top10 directors with the no:of movies and revenue where revenue should be taken into account for doing the analysis

Q16.Find the top10 directors with the no:of movies and revenue where movies should be taken into account for doing  the analysis

In [73]:
query = "SELECT directors.name, COUNT(original_title), SUM(revenue) FROM directors JOIN movies ON directors.id = movies.director_id GROUP BY director_id ORDER BY COUNT(original_title) DESC LIMIT 10"
cur.execute(query)
top_10 = cur.fetchall()
print(f'The top10 directors with the no:of movies and revenue are: {top_10}') 

The top10 directors with the no:of movies and revenue are: [('Steven Spielberg', 27, 9147393164), ('Woody Allen', 21, 669101038), ('Martin Scorsese', 20, 1956635998), ('Clint Eastwood', 20, 2512058888), ('Ridley Scott', 16, 3189557997), ('Robert Rodriguez', 16, 1109899581), ('Spike Lee', 16, 340618771), ('Steven Soderbergh', 15, 2114864443), ('Renny Harlin', 15, 1031500590), ('Tim Burton', 14, 3337418241)]


* Conclusion: Printed the top10 directors with the no:of movies and revenue where movies should be taken into account for doing the analysis

Q17. Give the title, release date, budget, revenue, popularity, and vote average made by Steven Spielberg

In [78]:
query = "SELECT original_title, release_date, budget, revenue, popularity, vote_average FROM movies JOIN directors ON movies.director_id = directors.id WHERE directors.name ='Steven Spielberg'"
cur.execute(query)
ss = cur.fetchall()
print(f'The title, release date, budget, revenue, popularity, and vote average made by Steven Spielberg is: {ss}')

The title, release date, budget, revenue, popularity, and vote average made by Steven Spielberg is: [('Indiana Jones and the Kingdom of the Crystal Skull', '2008-05-21', 185000000, 786636033, 75, 5.7), ('The BFG', '2016-06-01', 140000000, 183345589, 44, 6.0), ('War of the Worlds', '2005-06-28', 132000000, 591739379, 48, 6.2), ('The Adventures of Tintin', '2011-10-25', 130000000, 371940071, 89, 6.7), ('Minority Report', '2002-06-20', 102000000, 358372926, 65, 7.1), ('A.I. Artificial Intelligence', '2001-06-29', 100000000, 235926552, 34, 6.8), ('The Lost World: Jurassic Park', '1997-05-23', 73000000, 229074524, 2, 6.2), ('The Terminal', '2004-06-17', 60000000, 219417255, 57, 7.0), ('Munich', '2005-12-22', 70000000, 130358911, 29, 6.9), ('Hook', '1991-12-11', 70000000, 300854823, 33, 6.6), ('War Horse', '2011-12-25', 66000000, 177584879, 29, 7.0), ('Saving Private Ryan', '1998-07-24', 70000000, 481840909, 76, 7.9), ('Lincoln', '2012-11-09', 65000000, 275293450, 36, 6.7), ('Jurassic Park',

* Conclusion: The title, release date, budget, revenue, popularity, and vote average made by Steven Spielberg are printed

Q18. Give the total no:of films made my Steven Spielberg

In [77]:
query = "SELECT COUNT(original_title) FROM movies JOIN directors ON movies.director_id = directors.id WHERE directors.name = 'Steven Spielberg'"
cur.execute(query)
ss = cur.fetchall()
print(f'The no:of movies made by Steven Spielberg is: {ss[0]}')

The no:of movies made by Steven Spielberg is: (27,)


* Conclusion:  Steven Spielberg as directed 27 no:of movies